File: system-test/java/src/main/java/com/mycompany/myshop/testkit/dsl/core/usecase/external/clock/usecases/GetTime.java
package com.mycompany.myshop.testkit.dsl.core.usecase.external.clock.usecases;

import com.mycompany.myshop.testkit.driver.port.external.clock.ClockDriver;
import com.mycompany.myshop.testkit.driver.port.external.clock.dtos.GetTimeResponse;
import com.mycompany.myshop.testkit.dsl.core.usecase.external.clock.usecases.base.BaseClockUseCase;
import com.mycompany.myshop.testkit.dsl.core.shared.UseCaseResult;
import com.mycompany.myshop.testkit.dsl.core.shared.UseCaseContext;
import com.mycompany.myshop.testkit.driver.port.dtos.error.SystemError;

public class GetTime extends BaseClockUseCase<GetTimeResponse, GetTimeVerification> {
    public GetTime(ClockDriver driver, UseCaseContext context) {
        super(driver, context);
    }

    @Override
    public UseCaseResult<GetTimeResponse, GetTimeVerification> execute() {
        var result = driver.getTime();
        return new UseCaseResult<>(result.mapError(e -> SystemError.of(e.getMessage())), context, GetTimeVerification::new);
    }
}

File: system-test/typescript/src/testkit/dsl/core/usecase/external/clock/usecases/GetTime.ts
import type { ClockDriver } from '../../../../../../driver/port/external/clock/clock-driver.js';
import type { GetTimeResponse } from '../../../../../../driver/port/external/clock/dtos/GetTimeResponse.js';
import { success, failure, type Result } from '../../../../../../common/result.js';
import type { SystemError } from '../../../../../../driver/port/dtos/errors/SystemError.js';
import { UseCaseResult } from '../../../../shared/use-case-result.js';
import type { UseCaseContext } from '../../../../shared/use-case-context.js';
import { BaseClockUseCase } from './base/BaseClockUseCase.js';
import { GetTimeVerification } from './GetTimeVerification.js';

export class GetTime extends BaseClockUseCase<GetTimeResponse, GetTimeVerification> {
  constructor(driver: ClockDriver, context: UseCaseContext) {
    super(driver, context);
  }

  async execute(): Promise<UseCaseResult<GetTimeResponse, GetTimeVerification>> {
    const driverResult = await this.driver.getTime();
    const result: Result<GetTimeResponse, SystemError> = driverResult.success
      ? success(driverResult.value)
      : failure({ message: driverResult.error.message ?? 'Clock error', fieldErrors: [] });

    return new UseCaseResult(
      result,
      this.context,
      (response, ctx) => new GetTimeVerification(response, ctx),
    );
  }
}
